sharding/contracts/validator_manager.v.py

```diff
@@ -91,17 +115,24 @@ def withdraw(validator_index: num, sig: bytes <= 1000) -> bool:
     return result
 
 
-def sample(block_number: num, shard_id: num, sig_index: num) -> address:
+def sample(shard_id: num) -> address:
     zero_addr = 0x0000000000000000000000000000000000000000
 
-    cycle = floor(decimal(block_number / self.shuffling_cycle))
-    cycle_seed = blockhash(cycle * self.shuffling_cycle)
-    seed = blockhash(block_number)
-    index_in_subset = num256_mod(as_num256(sha3(concat(seed, as_bytes32(sig_index)))),
+    cycle = floor(decimal(block.number / self.shuffling_cycle_length))
+    cycle_start_block_number = cycle * self.shuffling_cycle_length - 1
+    if cycle_start_block_number < 0:
+        cycle_start_block_number = 0
+    cycle_seed = blockhash(cycle_start_block_number)
+    # originally, error occurs when block.number <= 4 because
+    # `seed_block_number` becomes negative in these cases.
+    # Now, just reject the cases when block.number <= 4
+    assert block.number >= self.period_length
+    seed = blockhash(block.number - (block.number % self.period_length) - 1)
+    index_in_subset = num256_mod(as_num256(sha3(concat(seed, as_bytes32(shard_id)))),
                                  as_num256(100))
```

Comment:
Should we use `self.shard_count` instead of `100`? Or it's just an arbitrary number?